diesel/src/macros/mod.rs

```diff
@@ -239,108 +263,235 @@ macro_rules! __diesel_column {
 /// ```
 #[macro_export]
 macro_rules! table {
-    // Put `use` statements at the end because macro_rules! cannot figure out
-    // if `use` is an ident or not (hint: It's not)
+    // Put imports into the import field
     (
-        use $($import:tt)::+; $($rest:tt)+
+        @parse
+        import = [$(use $($import:tt)::+;)*];
+        doc = [$($doc:meta,)*];
+        use $($new_import:tt)::+; $($rest:tt)+
     ) => {
-        table!($($rest)+ use $($import)::+;);
+        table! {
+            @parse
+            import = [$(use $($import)::+;)* use $($new_import)::+;];
+            doc = [$($doc,)*];
+            $($rest)+
+        }
     };
-
-    // Add the primary key if it's not present
+    // Put doc annotation into the doc field
     (
-        $($table_name:ident).+ {$($body:tt)*}
-        $($imports:tt)*
+        @parse
+        import = [$(use $($import:tt)::+;)*];
+        doc = [$($doc:meta,)*];
+        #[$new_doc:meta] $($rest:tt)+
     ) => {
         table! {
-            $($table_name).+ (id) {$($body)*} $($imports)*
+            @parse
+            import = [$(use $($import)::+;)*];
+            doc = [$($doc,)*$new_doc,];
+            $($rest)+
         }
     };
-
-    // Add the schema name if it's not present
+    // Forward to next parsing step
     (
-        $name:ident $(($($pk:ident),+))* {$($body:tt)*}
-        $($imports:tt)*
+        @parse
+        import = [$(use $($import:tt)::+;)+];
+        doc = [$($doc:meta,)*];
+        $($rest:tt)+
     ) => {
         table! {
-            public . $name $(($($pk),+))* {$($body)*} $($imports)*
+            @parse_body
+            import = [$(use $($import)::+;)*];
+            doc = [$($doc,)*];
+            $($rest)+
         }
     };
-
-    // Import `diesel::types::*` if no imports were given
+    // Forward to next parsing step and add default import (diesel::types::*)
     (
-        $($table_name:ident).+ $(($($pk:ident),+))* {$($body:tt)*}
+        @parse
+        import = [];
+        doc = [$($doc:meta,)*];
+        $($rest:tt)+
     ) => {
         table! {
-            $($table_name).+ $(($($pk),+))* {$($body)*}
-            use $crate::types::*;
+            @parse_body
+            import = [use $crate::types::*;];
+            doc = [$($doc,)*];
+            $($rest)+
+        }
+    };
+    // Add the primary key if it's not present
```

Comment:
Okay I can recognize the comments/sections but the code is quite a lot different. This will take me a while.